File: domain/src/test/java/cn/byronlab/weather/domain/usecase/DeleteSavedCityUseCaseTest.java
package cn.byronlab.weather.domain.usecase;

import cn.byronlab.weather.domain.model.City;
import cn.byronlab.weather.domain.model.Unit;
import cn.byronlab.weather.domain.repository.CityRepository;
import cn.byronlab.weather.domain.repository.SettingsRepository;
import cn.byronlab.weather.domain.result.DomainError;
import cn.byronlab.weather.domain.result.DomainResult;

import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class DeleteSavedCityUseCaseTest {

    @Test
    public void blankCityIdReturnsInvalidInput() {
        FakeCityRepository cityRepository = new FakeCityRepository();
        FakeSettingsRepository settingsRepository = new FakeSettingsRepository("101");
        DeleteSavedCityUseCase useCase = new DeleteSavedCityUseCase(cityRepository, settingsRepository);

        DomainResult<Unit> result = useCase.execute(" ");

        assertTrue(result.isFailure());
        assertEquals(DomainError.Type.INVALID_INPUT, result.getError().getType());
        assertEquals(0, cityRepository.deleteCallCount);
    }

    @Test
    public void deletingNonCurrentCityDoesNotChangeCurrentCity() {
        FakeCityRepository cityRepository = new FakeCityRepository(city("101"), city("102"));
        FakeSettingsRepository settingsRepository = new FakeSettingsRepository("102");
        DeleteSavedCityUseCase useCase = new DeleteSavedCityUseCase(cityRepository, settingsRepository);

        DomainResult<Unit> result = useCase.execute("101");

        assertTrue(result.isSuccess());
        assertEquals("102", settingsRepository.currentCityId);
        assertEquals(0, settingsRepository.setCallCount);
    }

    @Test
    public void deletingCurrentCitySwitchesToFirstRemainingCity() {
        FakeCityRepository cityRepository = new FakeCityRepository(city("101"), city("102"));
        FakeSettingsRepository settingsRepository = new FakeSettingsRepository("101");
        DeleteSavedCityUseCase useCase = new DeleteSavedCityUseCase(cityRepository, settingsRepository);

        DomainResult<Unit> result = useCase.execute("101");

        assertTrue(result.isSuccess());
        assertEquals("102", settingsRepository.currentCityId);
        assertEquals(1, settingsRepository.setCallCount);
    }

    @Test
    public void deletingLastCurrentCityClearsCurrentCity() {
        FakeCityRepository cityRepository = new FakeCityRepository(city("101"));
        FakeSettingsRepository settingsRepository = new FakeSettingsRepository("101");
        DeleteSavedCityUseCase useCase = new DeleteSavedCityUseCase(cityRepository, settingsRepository);

        DomainResult<Unit> result = useCase.execute("101");

        assertTrue(result.isSuccess());
        assertEquals("", settingsRepository.currentCityId);
    }

    private static City city(String cityId) {
        return new City(cityId, cityId, cityId, "", "", "", "");
    }

    private static final class FakeCityRepository implements CityRepository {

        private final List<City> savedCities = new ArrayList<>();
        private int deleteCallCount;

        private FakeCityRepository(City... cities) {
            savedCities.addAll(Arrays.asList(cities));
        }

        @Override
        public DomainResult<List<City>> searchCities(String keyword) {
            return DomainResult.success(new ArrayList<>(savedCities));
        }

        @Override
        public DomainResult<List<City>> getSavedCities() {
            return DomainResult.success(new ArrayList<>(savedCities));
        }

        @Override
        public DomainResult<Unit> deleteSavedCity(String cityId) {
            deleteCallCount++;
            for (int i = 0; i < savedCities.size(); i++) {
                if (cityId.equals(savedCities.get(i).getCityId())) {
                    savedCities.remove(i);
                    break;
                }
            }
            return DomainResult.success(Unit.INSTANCE);
        }
    }

    private static final class FakeSettingsRepository implements SettingsRepository {

        private String currentCityId;
        private int setCallCount;

        private FakeSettingsRepository(String currentCityId) {
            this.currentCityId = currentCityId;
        }

        @Override
        public DomainResult<String> getCurrentCityId() {
            return DomainResult.success(currentCityId);
        }

        @Override
        public DomainResult<Unit> setCurrentCityId(String cityId) {
            currentCityId = cityId;
            setCallCount++;
            return DomainResult.success(Unit.INSTANCE);
        }
    }
}

File: domain/src/test/java/cn/byronlab/weather/domain/usecase/DeleteSavedCityUseCaseTest.kt
package cn.byronlab.weather.domain.usecase

import cn.byronlab.weather.domain.model.City
import cn.byronlab.weather.domain.model.DeviceLocation
import cn.byronlab.weather.domain.model.WeatherRefreshInterval
import cn.byronlab.weather.domain.repository.CityRepository
import cn.byronlab.weather.domain.repository.SettingsRepository
import cn.byronlab.weather.domain.result.DomainError
import cn.byronlab.weather.domain.result.DomainResult
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class RemoveAddedCityUseCaseTest {

    @Test
    fun blankCityIdReturnsInvalidInput() = runTest {
        val cityRepository = FakeCityRepository()
        val settingsRepository = FakeSettingsRepository("101")
        val useCase = RemoveAddedCityUseCase(cityRepository, settingsRepository)

        val result = useCase(" ")

        assertTrue(result is DomainResult.Failure)
        assertEquals(DomainError.Type.INVALID_INPUT, (result as DomainResult.Failure).error.type)
        assertEquals(0, cityRepository.removeCallCount)
    }

    @Test
    fun deletingNonCurrentCityDoesNotChangeCurrentCity() = runTest {
        val cityRepository = FakeCityRepository(city("101"), city("102"))
        val settingsRepository = FakeSettingsRepository("102")
        val useCase = RemoveAddedCityUseCase(cityRepository, settingsRepository)

        val result = useCase("101")

        assertTrue(result is DomainResult.Success)
        assertEquals("102", settingsRepository.currentCityId)
        assertEquals(0, settingsRepository.setCallCount)
    }

    @Test
    fun deletingCurrentCitySwitchesToFirstRemainingCity() = runTest {
        val cityRepository = FakeCityRepository(city("101"), city("102"))
        val settingsRepository = FakeSettingsRepository("101")
        val useCase = RemoveAddedCityUseCase(cityRepository, settingsRepository)

        val result = useCase("101")

        assertTrue(result is DomainResult.Success)
        assertEquals("102", settingsRepository.currentCityId)
        assertEquals(1, settingsRepository.setCallCount)
    }

    @Test
    fun deletingLastCurrentCityClearsCurrentCity() = runTest {
        val cityRepository = FakeCityRepository(city("101"))
        val settingsRepository = FakeSettingsRepository("101")
        val useCase = RemoveAddedCityUseCase(cityRepository, settingsRepository)

        val result = useCase("101")

        assertTrue(result is DomainResult.Success)
        assertEquals("", settingsRepository.currentCityId)
    }

    private class FakeCityRepository(vararg cities: City) : CityRepository {

        private val addedCities = cities.toMutableList()
        var removeCallCount = 0

        override suspend fun searchCities(keyword: String): DomainResult<List<City>> {
            return DomainResult.success(addedCities.toList())
        }

        override suspend fun getPopularCities(): DomainResult<List<City>> {
            return DomainResult.success(addedCities.toList())
        }

        override fun resolveLocation(location: DeviceLocation): DomainResult<City> {
            return DomainResult.success(city("location"))
        }

        override fun observeAddedCities(): Flow<DomainResult<List<City>>> = flow {
            emit(getAddedCities())
        }

        override suspend fun getAddedCities(): DomainResult<List<City>> {
            return DomainResult.success(addedCities.toList())
        }

        override suspend fun addCity(cityId: String): DomainResult<Unit> {
            return DomainResult.success(Unit)
        }

        override suspend fun removeCity(cityId: String): DomainResult<Unit> {
            removeCallCount++
            addedCities.removeAll { it.cityId == cityId }
            return DomainResult.success(Unit)
        }

        override suspend fun getRecentCities() = DomainResult.success(emptyList<City>())

        override suspend fun recordRecentCity(cityId: String) = DomainResult.success(Unit)
    }

    private class FakeSettingsRepository(
        var currentCityId: String,
    ) : SettingsRepository {

        var setCallCount = 0

        override fun observeCurrentCityId(): Flow<DomainResult<String>> = flow {
            emit(getCurrentCityId())
        }

        override suspend fun getCurrentCityId(): DomainResult<String> {
            return DomainResult.success(currentCityId)
        }

        override suspend fun setCurrentCityId(cityId: String): DomainResult<Unit> {
            currentCityId = cityId
            setCallCount++
            return DomainResult.success(Unit)
        }

        override fun observeWeatherRefreshInterval(): Flow<DomainResult<WeatherRefreshInterval>> = flow {
            emit(getWeatherRefreshInterval())
        }

        override suspend fun getWeatherRefreshInterval() =
            DomainResult.success(WeatherRefreshInterval.ThirtyMinutes)

        override suspend fun setWeatherRefreshInterval(interval: WeatherRefreshInterval) =
            DomainResult.success(Unit)
    }
}

private fun city(cityId: String): City {
    return City(cityId, cityId, cityId, "", "", "", "")
}
